### ISSUE TITLE
The no-empty rule doesn't output anything if a comment is in the block

### ISSUE THREAD

**Issue by saadflow** [eslint/eslint#15006]
Hey y'all, I have submitted a PR to make comment behavior in the `no-empty` rule configurable (while keeping the current behavior as default). 
This is my first PR on this repo, I'd love to have some feedback 😁
Thanks !

**What rule do you want to change?**
`no-empty`
**Does this change cause the rule to produce more or fewer warnings?**
more
**How will the change be implemented? (New option, new default behavior, etc.)?**
A new option, whose default is the current behavior.
Currently, an empty block containing only comments doesn't produce any warning.
This is fine most of the time, but if for example you want to make sure that catch block don't contain // todo: handle error, it can be useful to make this behavior configurable.
This change adds an option to do just that, and if not provided, its default is to allow empty blocks with comments, so that the current behavior remains the default.
**Please provide some example code that this change will affect:**

```js
for(const foo of bar){
 // Keep me in the loop
}
```
```js
try {
   throwError()
} catch {
  // TODO: Handle this error.
}
```

**What does the rule currently do for this code?**
Currently it doesn't produce any error / warning.

**What will the rule do after it's changed?**
Now, it will do the same, unless the allowEmptyWithComments option is set to false, in which case it will produce an error / warning

**Are you willing to submit a pull request to implement this change?**
#14993 done

**Comment by saadflow** [comment]
Hi @mdjermanovic, thanks for your reply 😁

Agreed, I'm not sure either that this option would be widely used. But I still think it would be a great addition to eslint, here's why :

Every rule has its exceptions, valid cases. For every **other** rule in eslint, we use `eslint-disable` in order to enforce it throughout the project, and disable it only on the rare cases it's not relevant.
I don't think the `no-empty` should follow a different pattern and it feels inconsistent to "disable" the rule by writing a random comment who does not explicitly state that its purpose is to disable the rule.

Also, I don't see a lot of cases where an empty block would be valid. Especially `if` statements. Having `if(something){// do nothing }` is a clue that the logic could have been expressed differently, in more readable or efficient way. For instance and in my opinion, the second example you sent is missing a `return` statement in the first block to make it explicit that the function ends if the first expression is true.
This being said, most of the valid use cases are with catch block, like in the first example, and the rule already has an option to opt-out of it when the empty block is a catch block.

I can see how this is strongly opinionated 😁 but then again the option is totally opt-in. 

What's great about eslint is that no matter the opinion, there is a way to enforce it. Currently thought I can't make sure there are no code blocks that do not contain any logic.

**Comment by ljharb** [comment]
In try/catch, in particular, there's no way to swallow an exception from, say, JSON.parse, without an empty block. Requiring a comment in it to signal that it's intentional is identical to allowing extra parens in `if ((a = b))` to indicate that the assignment is intentional.

**Comment by mdjermanovic** [comment]
> Every rule has its exceptions, valid cases. For every **other** rule in eslint, we use `eslint-disable` in order to enforce it throughout the project, and disable it only on the rare cases it's not relevant.

I agree with this, but after thinking about this rule again I believe its purpose is more to enforce comments that explain why the block is empty than to really disallow empty blocks. That's why it's classified as a [`"suggestion"`](https://github.com/eslint/eslint/blob/4338b74767fa71e4e8d171f8503aa33d970e509f/lib/rules/no-empty.js#L19) rather than `"problem"`, and therefore adding this option might not make sense. (requiring that the comment matches a pattern might make sense, though)

In that regard, it's similar to the [default-case](https://eslint.org/docs/rules/default-case) rule - it doesn't really require that all switch statements have `default:`, but enforces a good practice to leave a comment if it was intentionally omitted.

If you want to disallow empty blocks without exceptions, that can be easily achieved with the [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax) rule:

```js
/* eslint no-restricted-syntax: 
    "error",
    {
        selector: ":not(FunctionDeclaration, FunctionExpression, ArrowFunctionExpression) > BlockStatement[body.length=0], SwitchStatement[cases.length=0]",
        message: "Unexpected empty block."
    }
]*/

if (foo) {
  // empty block, but this comment doesn't supress the error
}
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXG5cdFwiZXJyb3JcIixcbiAgICB7XG4gICAgICAgIHNlbGVjdG9yOiBcIjpub3QoRnVuY3Rpb25EZWNsYXJhdGlvbiwgRnVuY3Rpb25FeHByZXNzaW9uLCBBcnJvd0Z1bmN0aW9uRXhwcmVzc2lvbikgPiBCbG9ja1N0YXRlbWVudFtib2R5Lmxlbmd0aD0wXSwgU3dpdGNoU3RhdGVtZW50W2Nhc2VzLmxlbmd0aD0wXVwiLFxuICAgICAgICBtZXNzYWdlOiBcIlVuZXhwZWN0ZWQgZW1wdHkgYmxvY2suXCJcbiAgICB9XG5dKi9cblxuaWYgKGZvbykge1xuICAvLyBlbXB0eSBibG9jaywgYnV0IHRoaXMgY29tbWVudCBkb2Vzbid0IHN1cHJlc3MgdGhlIGVycm9yXG59XG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjMsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

This configuration disallows all empty blocks except for those covered by `no-empty-function` (and function bodies are not really blocks, but are represented as `BlockStatement` in the AST).

**Comment by nzakas** [comment]
I also don’t believe this makes sense to add a new option for. Between no-restricted-syntax and no-warning-comments, I think all of the bases are covered.

**Comment by saadflow** [comment]
> I believe its purpose is more to enforce comments that explain why the block is empty than to really disallow empty blocks.
If you look at it

**Comment by saadflow** [comment]
Yeah we understand the purpose of this rule in different ways 😁

> I believe its purpose is more to enforce comments that explain why the block is empty than to really disallow empty blocks.

Shouldn't it be called `document-empty` or `comment-empty` ? no means no, right ? 
Also,  its documentation clearly states : `This rule disallows empty block statements`, but the next line doesn't make much sense : 
>  This rule ignores block statements which contain a comment (for example, in an empty catch or finally block of a try statement to indicate that execution should continue regardless of errors).

The only example of a valid use case for an empty block that justifies using comments is actually covered by an existing option: `allowEmptyCatch`.

Don't get me wrong, a rule to enforce proper documentation of empty blocks would be neat, but I don't think this one's name is in sync with this purpose.

> In try/catch, in particular, there's no way to swallow an exception from, say, JSON.parse, without an empty block.

With or without comment, the `allowEmptyCatch` option allows this very much valid use case.

>If you want to disallow empty blocks without exceptions, that can be easily achieved with the no-restricted-syntax rule

Thanks 😁 ! I didn't know about it and this way I can still enforce no empty blocks if my PR isn't approved.

**Comment by nzakas** [comment]
To close the loop on this: the no-empty rule is intended to warn when you have an empty block. You can use a comment to opt-out because empty blocks may be valid, and you can document why that way. The rule considers the comment to indicate that the empty block was intentional and not a mistake. Without a comment, there’s just no way to distinguish those two cases.

As already mentioned, there’s not support on the team to move forward with this suggestion, so closing. You can, of course, create a custom rule to do exactly what you want or use the other options we’ve mentioned.
